### PR TITLE
Refactored RealmProcessor to make it easier to add changes.

### PR DESCRIPTION
**Pull request by cmelchior**
Part of #880 

This PR refactors RealmProcessor so all information about a ProxyClass is now generated and encapsulated inside ClassMetaData. This makes it easier to move information about the proxy classes around as well as making the responsibility of the RealmProcessor class much cleaner.

### CODE REVIEW

**Comment by cmelchior** [comment]
@emanuelez @kneth @bmunkholm

**Comment by kneth** [comment]
It's a lot of shuffling around but it seems fine to me.

File: realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+
+/**
+ * Utility class for holding metadata for RealmProxy classes.
+ */
+public class ClassMetaData {
+
+    private final TypeElement classType;
+    private String className;
+    private String packageName;
+    private boolean hasDefaultConstructor = false;
```
Comment (emanuelez) [inline realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java]: Redundant initialization

File: realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+    private VariableElement primaryKey = null;
```
Comment (emanuelez) [inline realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java]: Redundant initialization

File: realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+import io.realm.annotations.Ignore;
+import io.realm.annotations.Index;
+import io.realm.annotations.PrimaryKey;
+
+/**
+ * Utility class for holding metadata for RealmProxy classes.
+ */
+public class ClassMetaData {
+
+    private final TypeElement classType;
```
Comment (emanuelez) [inline realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java]: I would like to get this chance to document what all of these fields contain in detail

File: realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+        for (VariableElement field : fields) {
+            fieldNames.add(field.getSimpleName().toString());
+        }
+        for (VariableElement ignoredField : ignoredFields) {
+            fieldNames.add(ignoredField.getSimpleName().toString());
+            ignoreFieldNames.add(ignoredField.getSimpleName().toString());
+        }
+
+        for (ExecutableElement executableElement : methods) {
+
```
Comment (emanuelez) [inline realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java]: Remove this empty line

**Comment by emanuelez** [comment]
:+1: apart from comments

File: realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+    }
+
+    /**
+     * Build the meta data structures for this class. Any errors or messages will be
+     * posted on the provided Messager.
+     *
+     * @return True if processing should continue, false otherwise. Note that true might be returned for smaller errors,
+     * which will enable the annotation processor to output error messages for multiple classes before exiting.
+     */
+    public boolean generateMetaData(Messager messager) {
```
Comment (bmunkholm) [inline realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java]: This is still a HUGE method. Couldn't his be simplified for better readability? Better comments above each code section could at least help.

**Comment by cmelchior** [comment]
Ready for 2nd round after splitting huge method into smaller parts: @bmunkholm @emanuelez @kneth

**Comment by emanuelez** [comment]
:+1:

**Comment by bmunkholm** [comment]
Except as noted :+1:
Nice restructuring.

File: realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+
+        if (!found) {
+            Utils.note(String.format("Setter %s is not associated to any field", methodName));
+        }
+    }
+
+    // Verify that a getter is used to get a field in the model class.
+    // Note: This is done heuristically by comparing the name of getter with the name of the field.
+    // Annotation processors does not allow us to inspect individual statements.
+    private void checkGetterMethod(String methodName) {
```
Comment (kneth) [inline realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java]: When can this method be called? `expectedGetters` must exist before, right?

File: realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+        boolean found = false;
+
+        if (methodName.startsWith("is")) {
```
Comment (kneth) [inline realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java]: These methods should only exist for boolean fields. Isn't this the right place to check? I mean, the method is called `checkGetterMethod` but doesn't check the type.

File: realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+    }
+
+    // Report if the default constructor is missing
+    private void reportDefaultConstructorMissing() {
+        if (!hasDefaultConstructor) {
+            Utils.error("A default public constructor with no argument must be declared if a custom constructor is declared.");
+        }
+    }
+
+    // Loop through all class elements and add them to the appropriate internal data structures.
```
Comment (kneth) [inline realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java]: "Loop" -> "Iterate"

**Pull request merged**